compiler/rustc_trait_selection/src/solve/trait_goals.rs

```diff
@@ -120,6 +121,22 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
             return result;
         }
 
+        // Don't call `type_of` on a local TAIT that's in the defining
+        // scope, since that may require calling `typeck` on the same
+        // item we're currently type checking. Also don't do this when
+        // in `Reveal::All` mode, since we already normalize the self
+        // type in `assemble_candidates_after_normalizing_self_ty`.
+        if let ty::Alias(ty::Opaque, opaque_ty) = goal.predicate.self_ty().kind() {
+            if matches!(goal.param_env.reveal(), Reveal::All)
+                || opaque_ty
+                    .def_id
+                    .as_local()
+                    .is_some_and(|def_id| ecx.can_define_opaque_ty(def_id))
+            {
+                return Err(NoSolution);
```

Comment:
Can we have a comment about why this is fine in coherence?

Reply:
or, actually just generally. Is it because we handle the "normalized" form of the opaque elsewhere and that returns an ambiguous candidate?

Reply:
Clarified the comment above this block a bit more